static/js/narrow.js

```diff
@@ -287,79 +291,84 @@ exports.maybe_add_local_messages = function (opts) {
     var msg_data = opts.msg_data;
+    var unread_info;
 
-    if (select_strategy.flavor === 'first_unread') {
-        // Try to upgrade to the "exact" strategy by looking for
-        // unread message ids that match the upcoming narrow.
-        var unread_info = narrow_state.get_first_unread_info();
-
+    if (id_info.search_id) {
+        unread_info = narrow_state.get_first_unread_info();
         if (unread_info.flavor === 'found') {
-            // We found an unread message_id, but we won't return
-            // yet; we'll instead fall through to the next check
-            select_strategy = {
-                flavor: 'exact',
-                msg_id: unread_info.msg_id,
-            };
+            id_info.final_select_id = min_defined(
+                id_info.search_id,
+                unread_info.msg_id
+            );
         } else if (unread_info.flavor === 'not_found') {
-            // If we didn't find any unread messages, then we
-            // generally want to go the last message in the list,
-            // but only if we've fetched the newest messages
-            // and the narrow's not empty locally.
-            if (message_list.all.fetch_status.has_found_newest()) {
-                // Load messages now and upgrade our strategy
-                // if we find at least one message.
-                if (load_local_messages(msg_data)) {
-                    var last_msg = msg_data.last();
-                    select_strategy = {
-                        flavor: 'exact',
-                        msg_id: last_msg.id,
-                    };
-
-                    return select_strategy;
+            id_info.final_select_id = id_info.search_id;
+        } else {
+            // if we can't compute a next unread id, just
+            // return without setting local_select_id to be
+            // on the safe side
+            id_info.final_select_id = id_info.search_id;
+            return;
+        }
+
+        if (narrow_state.get_current_filter().can_apply_locally()) {
+            if (load_local_messages(msg_data)) {
+                if (msg_data.get(id_info.final_select_id)) {
+                    id_info.local_select_id = id_info.final_select_id;
                 }
             }
         }
+        return;
     }
 
-    // Do not make this an else-if...the block above could have
-    // changed select_strategy.
-    if (select_strategy.flavor === 'exact') {
-        var msg_to_select = message_list.all.get(select_strategy.msg_id);
-        if (msg_to_select !== undefined) {
+    unread_info = narrow_state.get_first_unread_info();
+
+    if (unread_info.flavor === 'found') {
+        id_info.final_select_id = min_defined(
+            id_info.previous_id,
+            unread_info.msg_id
+        );
+        if (load_local_messages(msg_data)) {
+            if (msg_data.get(id_info.final_select_id)) {
+                id_info.local_select_id = id_info.final_select_id;
+            }
+        }
+        return;
```

Comment:
The `found` case seems to be the same code here and in the `is_search` block; might make sense to pull it out of the `if search` block?

Reply:
Oh, I guess the difference is `previous_id` vs. `search_id`?